host/workflowidratelimit_test.go

```diff
@@ -0,0 +1,142 @@
+	suite.Run(t, s)
+}
+
+func (s *WorkflowIDRateLimitIntegrationSuite) SetupSuite() {
+	s.setupLogger()
+
+	s.Logger.Info("Running integration test against test cluster")
+	clusterMetadata := NewClusterMetadata(s.T(), s.testClusterConfig)
+	dc := persistence.DynamicConfiguration{
+		EnableCassandraAllConsistencyLevelDelete: dynamicconfig.GetBoolPropertyFn(true),
+		PersistenceSampleLoggingRate:             dynamicconfig.GetIntPropertyFn(100),
+		EnableShardIDMetrics:                     dynamicconfig.GetBoolPropertyFn(true),
+	}
+	params := pt.TestBaseParams{
+		DefaultTestCluster:    s.defaultTestCluster,
+		VisibilityTestCluster: s.visibilityTestCluster,
+		ClusterMetadata:       clusterMetadata,
+		DynamicConfiguration:  dc,
+	}
+	cluster, err := NewCluster(s.T(), s.testClusterConfig, s.Logger, params)
+	s.Require().NoError(err)
+	s.testCluster = cluster
+	s.engine = s.testCluster.GetFrontendClient()
+	s.adminClient = s.testCluster.GetAdminClient()
+
+	s.domainName = s.randomizeStr("integration-test-domain")
+	s.Require().NoError(s.registerDomain(s.domainName, 1, types.ArchivalStatusDisabled, "", types.ArchivalStatusDisabled, ""))
+
+	s.domainCacheRefresh()
+}
+
+func (s *WorkflowIDRateLimitIntegrationSuite) SetupTest() {
+	s.Assertions = require.New(s.T())
+}
+
+func (s *WorkflowIDRateLimitIntegrationSuite) TearDownSuite() {
+	s.tearDownSuite()
+}
+
+func (s *WorkflowIDRateLimitIntegrationSuite) TestWorkflowIDSpecificRateLimits() {
+	const (
+		testWorkflowID   = "integration-workflow-specific-rate-limit-test"
+		testWorkflowType = "integration-workflow-specific-rate-limit-test-type"
+		testTaskListName = "integration-workflow-specific-rate-limit-test-taskList"
+		testIdentity     = "worker1"
+	)
+
+	request := &types.StartWorkflowExecutionRequest{
+		RequestID:                           uuid.New(),
+		Domain:                              s.domainName,
+		WorkflowID:                          testWorkflowID,
+		WorkflowType:                        &types.WorkflowType{Name: testWorkflowType},
+		TaskList:                            &types.TaskList{Name: testTaskListName},
+		Input:                               nil,
+		ExecutionStartToCloseTimeoutSeconds: common.Int32Ptr(100),
+		TaskStartToCloseTimeoutSeconds:      common.Int32Ptr(1),
+		Identity:                            testIdentity,
+
+		WorkflowIDReusePolicy: types.WorkflowIDReusePolicyTerminateIfRunning.Ptr(),
+	}
+
+	ctx, cancel := createContext()
+	defer cancel()
+
+	// The ratelimit is 5 per second, so we should be able to start 5 workflows without any error
+	for i := 0; i < 5; i++ {
+		_, err := s.engine.StartWorkflowExecution(ctx, request)
+		assert.NoError(s.T(), err)
+	}
+
+	// Now we should get a rate limit error
+	for i := 0; i < 5; i++ {
+		_, err := s.engine.StartWorkflowExecution(ctx, request)
+		var busyErr *types.ServiceBusyError
+		assert.ErrorAs(s.T(), err, &busyErr)
+		assert.Equal(s.T(), common.WorkflowIDRateLimitReason, busyErr.Reason)
+	}
+}
```

Comment:
I wonder you do not test for sending the requests again after a second.
This would check for the whole cycle - accepting, ratelimiting, accepting again.

Reply:
Added a check in the end